projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/memory/LLVMStack.java

```diff
@@ -58,12 +60,13 @@ public LLVMStack(int stackSize) {
         lowerBounds = 0;
         upperBounds = 0;
         stackPointer = 0;
+        uniquesRegionPointer = 0;
         isAllocated = false;
     }
 
     public final class StackPointer implements AutoCloseable {
-        private long basePointer;
-        private long uniquesRegionBasePointer;
+        @CompilationFinal private long basePointer;
```

Comment:
I see that `CompilationFinal` is probably incorrect here due to different compilation units. I reverted this commit.

This commit is not directly related to uniques regions. It is just something I noticed when benchmarking the other optimizations. I found out that making this field `CompilationFinal` has a significant positive impact on the performance of Rust benchmarks on O0. In line with your investigation, I could only observe minor performance impacts on other Rust programs compiled on different optimization levels and on programs compiled with Clang.

I thought that all `StackPointer` objects of an `LLVMStack` except for the initial stack pointer should be constant, since successive stack pointers created via `newFrame` are never passed around ([example](https://github.com/graalvm/sulong/blob/71136c24081503691a82828a66b7f031bc170e76/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/func/LLVMDispatchNode.java#L90)). Maybe I have a misunderstanding here about what qualifies as "constant". 

In case you are interested, here is a benchmark which should reproduce the significant performance gain I observed: [fannkuchbench-n10.rs](https://gist.github.com/anatol1234/184190866062b66940b21cac1034aa77)

Commandline (Rust stable 1.19.0) - first iteration takes about a minute:

    rustc --emit=llvm-bc fannkuchbench-n10.rs
    mx --dynamicimport /compiler --jdk jvmci lli -Dpolyglot.llvm.libraries=<Rust stdlib for 1.19.0>  fannkuchbench-n10.bc

After about 3 iterations you should see a speedup of about 40% compared to a Sulong build without this commit. However, given that `CompilationFinal` is incorrect here, I can imagine that as a consequence some incorrect assumptions are made about the stack pointers  which speed up the execution in this case, but could lead to memory violations or incorrect results in other scenarios.

Reply:
`newFrame` allocates a new object, so the result is never a compile-time constant. It is a different value at every call.

But the `StackPointer` objects allocated with `newFrame` should be escape-analyzed, that is, the object is never really allocated, and therefore the value of the `basePointer` field is not really a field read, but directly connected to the value that was originally written. In that case, `@CompilationFinal` or `final` should not make any difference at all, only the constant-ness of the original value is used.

That's the reason I'm confused why it can make any performance difference at all.

I will investigate further. I'll merge the rest of this PR in the meantime.